cmd/peer-rest-server.go

```diff
@@ -547,6 +549,38 @@ func (s *peerRESTServer) SignalServiceHandler(w http.ResponseWriter, r *http.Req
 	}
 }
 
+func (s *peerRESTServer) TraceHandler(w http.ResponseWriter, r *http.Request) {
+	trcAll := false
+	if a := r.URL.Query().Get("all"); a != "false" {
+		trcAll = true
+	}
+	targetID := r.URL.Query().Get("target-id")
+	ch := globalTrace.pubsub.Subscribe()
+	defer globalTrace.pubsub.Unsubscribe(ch)
+	globalTrace.traceTargets[targetID] = ch
+	for entry := range ch {
+		trcInfo := entry.(trace.Info)
+		// omit inter-node traffic if trcAll is false
+		if !trcAll && strings.HasPrefix(trcInfo.ReqInfo.URL.Path, "/minio") {
+			continue
+		}
+		data, err := json.Marshal(trcInfo)
+		if err != nil {
+			return
+		}
+		data = append(data, byte('\n'))
+		if _, err = w.Write(data); err != nil {
+			return
+		}
+		w.(http.Flusher).Flush()
+	}
+}
```

Comment:
You can avoid out of band Unsubscribe rest call by doing this
```
func (s *peerRESTServer) TraceHandler(w http.ResponseWriter, r *http.Request) {
        trcAll := false
        if a := r.URL.Query().Get("all"); a != "false" {
                trcAll = true
        }
        doneCh := make(chan struct{})
        go func() {
                io.Copy(ioutil.Discard, r.Body)
                close(doneCh)
        }()
        ch := globalTrace.pubsub.Subscribe()
        defer globalTrace.pubsub.Unsubscribe(ch)

        for {
                select {
                case <-doneCh:
                        return
                case entry := <-ch:
                        trcInfo := entry.(trace.Info)
                        // omit inter-node traffic if trcAll is false
                        if !trcAll && strings.HasPrefix(trcInfo.ReqInfo.URL.Path, "/minio") {
                                continue
                        }
                        data, err := json.Marshal(trcInfo)
                        if err != nil {
                                return
                        }
                        data = append(data, byte('\n'))
                        if _, err = w.Write(data); err != nil {
                                return
                        }
                        w.(http.Flusher).Flush()
                }
        }
}
```

Reply:
Also in the above for loop, you can select on `GlobalServiceDoneCh` and remove the function stopHTTPTrace